beetsplug/embedart.py

```diff
@@ -149,7 +149,7 @@ def embed_func(lib, opts, args):
                     with open(tempimg, "wb") as f:
                         f.write(response.content)
                 except Exception as e:
-                    self._log.error("Unable to save image: {}".format(e))
+                    self._log.error(f"Unable to save image: {e}")
```

Comment:
Since we don't have any benchmarking tests at the moment, I'm not sure how much of a difference this would be. Since this is largely a disk-bound program and we use multiple threads, I think having fstrings through the code as standard might be a better advantage than whatever mild speed gains there are from this.

The pylint page says that using f-strings is a reasonable option. 

Plus we're not passing through to the logging module transparently, so the string interpolation is being done anyway from the looks of things.  If we change that module, then it might make sense to switch just the logging events, but even then, we'd need benchmarks to see.

Reply:
My concern has less to do with speed and more with errors.

For example, I edited `FieldSort` implementation to embed an error into its `__repr__` method:

```python
class FieldSort(Sort):
    ...
    def __repr__(self) -> str:
        return (
            f"{self.__class__.__name__}{1 / 0}"
            f"({repr(self.field)}, ascending={repr(self.ascending)})"
        )
```

I'm using beets logger. Note that the logging level is `30` (WARNING)

```python
from beets.logging import getLogger

log = getLogger(__name__)
log.getEffectiveLevel()
# 30

s = FieldSort("field")
```

#### Using f-string

```python
[ins] In [20]: log.debug(f"Sort {s}")
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ in <module>:1                                                                                    │
│                                                                                                  │
│ /home/sarunas/repo/beets/beets/dbcore/query.py:1019 in __repr__                                  │
│                                                                                                  │
│   1016 │                                                                                         │
│   1017 │   def __repr__(self) -> str:                                                            │
│   1018 │   │   return (                                                                          │
│ ❱ 1019 │   │   │   f"{self.__class__.__name__}{1 / 0}"                                           │
│   1020 │   │   │   f"({repr(self.field)}, ascending={repr(self.ascending)})"                     │
│   1021 │   │   )                                                                                 │
│   1022                                                                                           │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
ZeroDivisionError: division by zero
```

#### Lazy logging
```python
log.debug("Sort: {}", s)
```

With an f-string this issue is raised immediately in the main thread, regardless of whether the user runs beets in verbose mode or not. With lazy logging, that logic does not run.

Lazily, even when this is allowed to be logged out, there's no exception raised in the main logic:

```python
[ins] In [30]: log.warning("Sort: {}", s)
          ...: print("\n---\nMain logic continues")
--- Logging error ---
Traceback (most recent call last):
  File "/home/sarunas/.local/share/pyenv/versions/3.8.19/lib/python3.8/logging/__init__.py", line 1085, in emit
    msg = self.format(record)
  File "/home/sarunas/.local/share/pyenv/versions/3.8.19/lib/python3.8/logging/__init__.py", line 929, in format
    return fmt.format(record)
  File "/home/sarunas/.local/share/pyenv/versions/3.8.19/lib/python3.8/logging/__init__.py", line 668, in format
    record.message = record.getMessage()
  File "/home/sarunas/.local/share/pyenv/versions/3.8.19/lib/python3.8/logging/__init__.py", line 371, in getMessage
    msg = str(self.msg)
  File "/home/sarunas/repo/beets/beets/logging.py", line 70, in __str__
    return self.msg.format(*args, **kwargs)
  File "/home/sarunas/repo/beets/beets/dbcore/query.py", line 1019, in __repr__
    f"{self.__class__.__name__}{1 / 0}"
ZeroDivisionError: division by zero
Call stack:
  File "/media/poetry/virtualenvs/beets-yAypcYUQ-py3.8/bin/ptipython", line 8, in <module>
    sys.exit(run())
  File "/media/poetry/virtualenvs/beets-yAypcYUQ-py3.8/lib/python3.8/site-packages/ptpython/entry_points/run_ptipython.py", line 72, in run
    embed(
  File "/media/poetry/virtualenvs/beets-yAypcYUQ-py3.8/lib/python3.8/site-packages/ptpython/ipython.py", line 323, in embed
    shell(header=header, stack_depth=2, compile_flags=compile_flags)
  File "/media/poetry/virtualenvs/beets-yAypcYUQ-py3.8/lib/python3.8/site-packages/IPython/terminal/embed.py", line 251, in __call__
    self.mainloop(
  File "/media/poetry/virtualenvs/beets-yAypcYUQ-py3.8/lib/python3.8/site-packages/IPython/terminal/embed.py", line 343, in mainloop
    self.interact()
  File "/media/poetry/virtualenvs/beets-yAypcYUQ-py3.8/lib/python3.8/site-packages/IPython/terminal/interactiveshell.py", line 881, in interact
    self.run_cell(code, store_history=True)
  File "/media/poetry/virtualenvs/beets-yAypcYUQ-py3.8/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3009, in run_cell
    result = self._run_cell(
  File "/media/poetry/virtualenvs/beets-yAypcYUQ-py3.8/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3064, in _run_cell
    result = runner(coro)
  File "/media/poetry/virtualenvs/beets-yAypcYUQ-py3.8/lib/python3.8/site-packages/IPython/core/async_helpers.py", line 129, in _pseudo_sync_runner
    coro.send(None)
  File "/media/poetry/virtualenvs/beets-yAypcYUQ-py3.8/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3269, in run_cell_async
    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,
  File "/media/poetry/virtualenvs/beets-yAypcYUQ-py3.8/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3448, in run_ast_nodes
    if await self.run_code(code, result, async_=asy):
  File "/media/poetry/virtualenvs/beets-yAypcYUQ-py3.8/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3508, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-30-ecad3783550f>", line 1, in <module>
    log.warning("Sort: {}", s)
  File "/home/sarunas/.local/share/pyenv/versions/3.8.19/lib/python3.8/logging/__init__.py", line 1458, in warning
    self._log(WARNING, msg, args, **kwargs)
  File "/home/sarunas/repo/beets/beets/logging.py", line 88, in _log
    return super()._log(
Message: <beets.logging.StrFormatLogger._LogMessage object at 0x765899d57310>
Arguments: ()

---
Main logic continues
```

Reply:
I disagree with this argument.  It seems impossible to me that any of the formatting logic in the codebase can fail depending on the current logging level.  Any f-string passed to `log` should be safe to evaluate at any logging level.  If you can find an actual example of this not being true in the current codebase, please show us.

Reply:
> I disagree with this argument. It seems impossible to me that any of the formatting logic in the codebase can fail depending on the current logging level.

Sorry, I must have miscommunicated - normally, when logging logic is clean and does not raise any exceptions, there is no difference between f-string and lazy formatting (except for a slight performance difference). 

On the other hand, if logging logic happens to be broken, for example, an exception is raised in `__repr__` method,

* if the logging message is formatted with an **f-string**
  ```python
  self._log.debug(f"Object: {repr(obj)}")
  ```
  Since `obj.__repr__` runs when this line is executed, the exception is **always** raised immediately (regardless of the logging level) in the main logic. If this is not handled, Python exits immediately. That's because Python executes an f-string expression right when it comes across one.
* using **lazy logging**
  ```python
  self._log.debug("Object: {}", obj)
  ```
  `obj.__repr__` is called lazily if the user's configured logging level allows the message to be logged out. If `obj.__repr__` fails, the exception is only printed in the background and **never** raised in the main logic. That's because the broken expression is not in the source code (Python only comes across `obj` when it reads this line). Instead, `obj.__repr__` gets executed by the `logging` module following a form of this (simplified) logic:
  ```python
  def debug(self):
      if USER_LOGGING_LEVEL >= DEBUG:
          try:
              # note __repr__ gets called implicitly if an object does not have __str__ method
              print(self.msg.format(self.args))
          except Exception as e:
              print(e)
  ``` 

* Note that in the example in my previous comment, I adjusted a `__repr__` method implementation with `1 / 0` expression in order to simulate this issue.

```python
    def __repr__(self) -> str:
        return (
            f"{self.__class__.__name__}{1 / 0}"
            f"({repr(self.field)}, ascending={repr(self.ascending)})"
        )
```

Ultimately, none of this is based on my subjective personal opinion - this is one of Python's logging best practices. If you're interested, have a read [here](https://github.com/pylint-dev/pylint/issues/2395)

Reply:
Well ... there _shouldn't_ be any errors in the logging logic.  If there are, we want to find out about them and fix them.  Making them less noticeable doesn't really help that.  We _want_ users to notice that something is wrong with the program, because there otherwise we can't find out that there is something wrong.

Reply:
> If there are, we want to find out about them and fix them.

We've got 62% coverage, so we need to test the rest of 38% of code in order to be sure about this.

> Making them less noticeable doesn't really help that. We want users to notice that something is wrong with the program, because there otherwise we can't find out that there is something wrong.

I agree that visibility is important. But are we ready to accept that this crashes beets immediately, since an error in an f-string is not handled?

Reply:
> Even if a crash or two does occur, I think it's worth it.

I don't think I am on the same page regarding this. Mistakes happen, and issues slip through the reviews (https://github.com/beetbox/beets/pull/5356). If I accidentally merge some untested code which raises an exception while formatting `log.debug` message, this will make beets unusable until the issue is addressed and the fix is merged.

Some proactive users will report the issue here, however others (especially new) users will stop using beets since it's not functional.

At least for me, this drawback is more important than anything we may gain from increased visibility.

Even if we are happy to accept the above, most of the code base uses lazy logging, so I think we want to be consistent with what we already have.

For example, we've got 28 `log.debug` statements that use non-lazy logging:

```
$ oneline_python **/*.py | grep -E 'log.debug\(.*(format\(|f"|" \+ )'
            log.debug("found duplicates: {}".format([o.id for o in found_duplicates]))
                self._log.debug(f"adding simple rewrite '{pattern}' → '{value}' " f"for field {fieldname}")
                    self._log.debug(f"adding advanced rewrite to '{replacement}' " f"for field {fieldname}")
        log.debug("Listening for control signals on {}:{}".format(host, ctrl_port))
        self._log.debug(f"Searching {self.data_source} for '{query}'")
                self._log.debug("{}: error receiving response".format(self.NAME))
                self._log.debug("{}: error loading response: {}".format(self.NAME, response.text))
            self._log.debug("google: error loading response: {}".format(response.text))
            self._log.debug("lastfm: error loading response: {}".format(response.text))
            self._log.debug("Error: " + str(e))
            self._log.debug("Spotify: error loading response: {}".format(response.text))
            self._log.debug(f"Cover art URL {image_url} found for {album}")
            self._log.debug(f"Cover art URL not found for {album}")
            self._log.debug(f"Cover art URL not found for {album}")
                        self._log.debug("using {0.LOC_STR} image {1}".format(source, util.displayable_path(out.path)))
            self._log.debug(f"Invalid Search Error: {e}")
        self._log.debug(f"{self.__class__.__name__}: {message}", *args)
            self._log.debug(f"loading iTunes library from {library_path}")
        self._log.debug(f"analyzing {item}")
            self._log.debug("{}: {} blocks over {} LUFS".format(item, n_blocks, gating_threshold))
        self._log.debug("{}: gain {} LU, peak {}".format(item, gain, peak))
                self._log.debug(f"{self.data_source} access token has expired. " f"Reauthenticating.")
                self._log.debug(f"Too many API requests. Retrying after " f"{seconds} seconds.")
        self._log.debug(f"Searching {self.data_source} for '{query}'")
        self._log.debug(f"using {ArtResizer.shared.method} to write metadata")
                log.debug(f"artresizer: method is {self.local_method.NAME}")
                self._log.debug("debug " + name)
            self._log.debug("debug " + name)
```

And 237 that use lazy logging:

```
$ oneline_python **/*.py | grep 'log.debug(.*{' | grep -vE 'log.debug.*(format\(|f"|" \+ )'
        log.debug("embedding {0}", displayable_path(imagepath))
    log.debug("Resizing album art to {0} pixels wide and encoding at quality \
        log.debug("Clearing art for {0}", item)
    log.debug("Searching for discovered album ID: {0}", first)
    log.debug("Candidate: {0} - {1} ({2})", info.artist, info.album, info.album_id)
            log.debug("Ignored. Missing required tag: {0}", req_tag)
            log.debug("Ignored. Penalty: {0}", penalty)
    log.debug("Success. Distance: {0}", dist)
    log.debug("Tagging {0} - {1}", cur_artist, cur_album)
            log.debug("Searching for album ID: {0}", search_id)
            log.debug("Album ID match recommendation is {0}", rec)
        log.debug("Search terms: {0} - {1}", search_artist, search_album)
            log.debug("Additional search terms: {0}", extra_tags)
        log.debug("Album might be VA: {0}", va_likely)
    log.debug("Evaluating {0} candidates.", len(candidates))
            log.debug("Searching for track ID: {0}", trackid)
    log.debug("Item search terms: {0} - {1}", search_artist, search_title)
    log.debug("Found {0} candidates.", len(candidates))
        log.debug("Album {} has too many tracks", release["id"])
            log.debug("Retrieving tracks starting at {}", i)
                log.debug("Found link to {} release via MusicBrainz", source.capitalize())
        log.debug("Searching for MusicBrainz releases with: {!r}", criteria)
    log.debug("Requesting MusicBrainz release {}", releaseid)
        log.debug("Invalid MBID ({0}).", releaseid)
        log.debug("Invalid MBID ({0}).", releaseid)
        log.debug("state file could not be read: {0}", exc)
        log.debug("removing {0} old duplicated items", len(duplicate_items))
                log.debug("deleting duplicate {0}", util.displayable_path(item.path))
            log.debug("Set field {1}={2} for {0}", displayable_path(self.paths), field, value)
                log.debug("Reimported {} {}. Not preserving flexible attributes {}. " "Path: {}", noun, new_obj.id, overwritten_fields, displayable_path(new_obj.path))
                log.debug("Reimported album {}. Preserving attribute ['added']. " "Path: {}", self.album.id, displayable_path(self.album.path))
                log.debug("Reimported album {}. Preserving flexible attributes {}. " "Path: {}", self.album.id, list(album_fields.keys()), displayable_path(self.album.path))
                    log.debug("Reimported item {}. Preserving attribute ['added']. " "Path: {}", item.id, displayable_path(item.path))
                log.debug("Reimported item {}. Preserving flexible attributes {}. " "Path: {}", item.id, list(item_fields.keys()), displayable_path(item.path))
                log.debug("Replacing item {0}: {1}", dup_item.id, displayable_path(item.path))
        log.debug("{0} of {1} items replaced", sum(bool(v) for v in self.replaced_items.values()), len(self.imported_items()))
            log.debug("Set field {1}={2} for {0}", displayable_path(self.paths), field, value)
            log.debug("Removing extracted directory: {0}", displayable_path(self.toppath))
            log.debug("Skipping previously-imported path: {0}", displayable_path(path))
            log.debug("Skipping previously-imported path: {0}", displayable_path(dirs))
        log.debug("Extracting archive: {0}", displayable_path(self.toppath))
        log.debug("Archive extracted to: {0}", self.toppath)
            log.debug("yielding album {0}: {1} - {2}", album.id, album.albumartist, album.album)
    log.debug("Looking up: {0}", displayable_path(task.paths))
            log.debug("default action for duplicates: {0}", duplicate_action)
                log.debug("moving {0} to synchronize path", util.displayable_path(self.path))
        log.debug("moving album art {0} to {1}", util.displayable_path(old_art), util.displayable_path(new_art))
    log.debug("Parsed query: {!r}", query)
    log.debug("Parsed sort: {!r}", sort)
            self._log.debug("Successfully submitted AcousticBrainz analysis " "for {}.", item)
            self._log.debug("fetching URL: {}", url)
                self._log.debug("Invalid Response: {}", res.text)
                        self._log.debug("attribute {} of {} set to {}", attr, item, val)
                        self._log.debug("skipping attribute {} of {}" " (value {}) due to config", attr, item, val)
                self._log.debug("Data {} could not be mapped to scheme {} " "because key {} was not found", subdata, v, k)
        self._log.debug("running command: {}", displayable_path(list2cmdline(cmd)))
        self._log.debug("checking path: {}", dpath)
            self._log.debug("authentication error: {0}", e)
            self._log.debug("authentication error: {0}", e)
        self._log.debug("Beatport token {0}, secret {1}", token, secret)
            self._log.debug("API Error: {0} (query: {1})", e, query)
            self._log.debug("API Error: {0} (query: {1})", e, query)
        self._log.debug("Searching for release {0}", release_id)
        self._log.debug("Searching for track {0}", track_id)
        self.server._log.debug("{}[{}]: {}", kind, self.address, message)
        self.server._log.debug("CTRL {}[{}]: {}", kind, self.address, message)
                    self._log.debug("moving album {}", album)
        log.debug("fingerprint matching {0} failed: {1}", util.displayable_path(repr(path)), exc)
    log.debug("chroma: fingerprinted {0}", util.displayable_path(repr(path)))
    log.debug("matched recordings {0} on releases {1}", recording_ids, release_ids)
        self._log.debug("acoustid album candidates: {0}", len(albums))
        self._log.debug("acoustid item candidates: {0}", len(tracks))
            self._log.debug("Command {0} exited with status {1}: {2}", args, exc.returncode, exc.output)
                    self._log.debug("embedding album art from {}", util.displayable_path(album.artpath))
            self._log.debug("image size: {}", size)
            self._log.debug("Error fetching album tracks for {}", deezer_id)
            self._log.debug("Error fetching album tracks for {}", track_data["album"]["id"])
        self._log.debug("Found {} result(s) from {} for '{}'", len(response_data), self.data_source, query)
                self._log.debug("No deezer_track_id present for: {}", item)
                self._log.debug("Deezer track: {} has {} rank", deezer_track_id, rank)
                self._log.debug("Invalid Deezer track_id: {}", e)
            self._log.debug("connection error: {0}", e)
            self._log.debug("connection error: {0}", e)
        self._log.debug("Discogs token {0}, secret {1}", token, secret)
            self._log.debug("API Error: {0} (query: {1})", e, query)
            self._log.debug("API Error: {0} (query: {1})", e, query)
            self._log.debug("searching within album {0}", album_cur.album)
        self._log.debug("Searching for release {0}", album_id)
                self._log.debug("API Error: {0} (query: {1})", e, result.data["resource_url"])
            self._log.debug("Communication error while searching for {0!r}", query, exc_info=True)
        self._log.debug("Searching for master release {0}", master_id)
                self._log.debug("API Error: {0} (query: {1})", e, result.data["resource_url"])
            self._log.debug("{}", traceback.format_exc())
            self._log.debug("Invalid position: {0}", position)
            self._log.debug("key {0} on item {1} not cached:" "computing checksum", key, displayable_path(item.path))
                self._log.debug("computed checksum for {0} using {1}", item.title, key)
                self._log.debug("failed to checksum {0}: {1}", displayable_path(item.path), e)
            self._log.debug("key {0} on item {1} cached:" "not computing checksum", key, displayable_path(item.path))
                self._log.debug("some keys {0} on item {1} are null or empty:" " skipping", keys, displayable_path(obj.path))
                self._log.debug("all keys {0} on item {1} are null or empty:" " skipping", keys, displayable_path(obj.path))
                        self._log.debug("key {0} on item {1} is null " "or empty: setting from item {2}", f, displayable_path(objs[0].path), displayable_path(o.path))
                    self._log.debug("item {0} missing from album {1}:" " merging from {2} into {3}", missing, objs[0], displayable_path(o.path), displayable_path(missing.destination()))
    log.debug("invoking editor command: {!r}", cmd)
                self._log.debug("saving changes to {}", ob)
                self._log.debug("Removing album art file for {0}", album)
        self._log.debug("image size: {}", self.size)
            self._log.debug("image too small ({} < {})", self.size[0], plugin.minwidth)
                    self._log.debug("image is not close enough to being " "square, ({} - {} > {})", long_edge, short_edge, plugin.margin_px)
                    self._log.debug("image is not close enough to being " "square, ({} - {} > {})", long_edge, short_edge, margin_px)
                self._log.debug("image is not square ({} != {})", self.size[0], self.size[1])
            self._log.debug("image needs rescaling ({} > {})", self.size[0], plugin.maxwidth)
                self._log.debug("image needs resizing ({}B > {}B)", filesize, plugin.max_filesize)
                self._log.debug("image needs reformatting: {} -> {}", fmt, plugin.cover_format)
        log.debug("{}: {}", message, prepped.url)
                    self._log.debug("not a supported image: {}", real_ct or "unknown content type")
                self._log.debug("downloaded art to: {0}", util.displayable_path(filename))
            self._log.debug("error fetching art: {}", exc)
                self._log.debug("error cleaning up tmp art: {}", exc)
            self._log.debug("scraped art URL: {0}", resp.url)
            self._log.debug("google fetchart error: {0}", reason)
            self._log.debug("fanart.tv: error loading response: {}", response.text)
                self._log.debug("fanart.tv: error on request: {}", data["error message"])
            self._log.debug("iTunes search failed: {0}", e)
            self._log.debug("Could not decode json response: {0}", e)
            self._log.debug("{} not found in json. Fields are {} ", e, list(r.json().keys()))
            self._log.debug("iTunes search for {!r} got no results", payload["term"])
                self._log.debug("Malformed itunes candidate: {} not found in {}", e, list(c.keys()))
            self._log.debug("Malformed itunes candidate: {} not found in {}", e, list(c.keys()))
            self._log.debug("wikipedia: error scraping dbpedia response: {}", dbpedia_response.text)
                    self._log.debug("using well-named art file {0}", util.displayable_path(fn))
                self._log.debug("using fallback art file {0}", util.displayable_path(remaining[0]))
                    self._log.debug("lastfm: no results for {}", album.mb_albumid)
            self._log.debug("Storing art_source for {0.albumartist} - {0.album}", album)
                self._log.debug("trying source {0} for album {1.albumartist} - {1.album}", SOURCE_NAMES[type(source)], album)
            self._log.debug('running command "{0}" for event {1}', " ".join(command_pieces), event)
        self._log.debug("Recorded mtime {0} for item '{1}' imported from " "'{2}'", mtime, util.displayable_path(destination), util.displayable_path(source))
            self._log.debug("Album '{0}' is reimported, skipping import of " "added dates for the album and its items.", util.displayable_path(album.path))
        self._log.debug("Import of album '{0}', selected album.added={1} " "from item file mtimes.", album.album, album.added)
            self._log.debug("Item '{0}' is reimported, skipping import of " "added date.", util.displayable_path(item.path))
            self._log.debug("Import of item '{0}', selected item.added={1}", util.displayable_path(item.path), item.added)
            self._log.debug("Write of item '{0}', selected item.added={1}", util.displayable_path(item.path), item.added)
            self._log.debug("adding item field {0}", key)
            self._log.debug("adding album field {0}", key)
    log.debug("Sending event: {0}", event)
        log.debug("Extracting {} ID from '{}'", url_type, id_)
                self._log.debug("{0} already added", album_dir)
            self._log.debug("Loading canonicalization tree {0}", c14n_filename)
                    self._log.debug("added last.fm item genre ({0}): {1}", src, item.genre)
            self._log.debug("added last.fm album genre ({0}): {1}", src, album.genre)
                    self._log.debug("added last.fm item genre ({0}): {1}", src, item.genre)
            self._log.debug("added last.fm item genre ({0}): {1}", src, item.genre)
            self._log.debug("last.fm error: {0}", exc)
            self._log.debug("{}", traceback.format_exc())
        log.debug("query: {0} - {1} ({2})", artist, title, album)
            log.debug("no album match, trying by album/title: {0} - {1}", album, title)
            log.debug("match: {0} - {1} ({2}) " "updating: play_count {3} => {4}", song.artist, song.title, song.album, count, new_count)
            self._log.debug("loading extension {}", ext)
            self._log.debug("applying changes to {}", album_formatted)
                        self._log.debug("moving album {0}", album_formatted)
                    self._log.debug("track {0} in album {1}", track_info.track_id, album_info.album_id)
        self._log.debug("music_directory: {0}", self.music_directory)
        self._log.debug("strip_path: {0}", self.strip_path)
        self._log.debug("returning: {0}", result)
            self._log.debug("updated: {0} = {1} [{2}]", attribute, item[attribute], displayable_path(item.path))
                    self._log.debug('unhandled status "{0}"', status)
            self._log.debug("no composer for {}; add one at " "https://musicbrainz.org/work/{}", item, work_info["work"]["id"])
                self._log.debug("error fetching work: {}", e)
                self._log.debug("Work fetched: {} - {}", parent_info["parentwork"], parent_info["parent_composer"])
                self._log.debug("Work fetched: {} - no parent composer", parent_info["parentwork"])
            self._log.debug("{}: Work present, skipping", item)
        log.debug("set permissions to {}, but permissions are now {}", permission, os.stat(syspath(path)).st_mode & 0o777)
            self._log.debug("setting file permissions on {}", displayable_path(path))
            self._log.debug("setting directory permissions on {}", displayable_path(path))
    log.debug("executing command: {} {!r}", command_str, open_args)
        self._log.debug("applied track gain {0} LU, peak {1} of FS", item.rg_track_gain, item.rg_track_peak)
        self._log.debug("applied album gain {0} LU, peak {1} of FS", item.rg_album_gain, item.rg_album_peak)
        self._log.debug("done analyzing {0}", item)
            self._log.debug("done analyzing {0}", item)
        self._log.debug("applied r128 track gain {0} LU", item.r128_track_gain)
        self._log.debug("applied r128 album gain {0} LU", item.r128_album_gain)
        self._log.debug("{}: gain {} LU, peak {}", task.album, album_gain, album_peak)
        self._log.debug("executing {0}", " ".join(map(displayable_path, cmd)))
        self._log.debug("analyzing {0} files", len(items))
        self._log.debug("executing {0}", " ".join(map(displayable_path, cmd)))
                self._log.debug("bad tool output: {0}", text)
            self._log.debug("error in rg.title_gain() call: {}", exc)
        self._log.debug("ReplayGain for track {0} - {1}: {2:.2f}, {3:.2f}", item.artist, item.title, rg_track_gain, rg_track_peak)
            self._log.debug("ReplayGain for track {0}: {1:.2f}, {2:.2f}", item, rg_track_gain, rg_track_peak)
        self._log.debug("ReplayGain for album {0}: {1:.2f}, {2:.2f}", task.items[0].album, rg_album_gain, rg_album_peak)
            self._log.debug("adding template field {0}", key)
            self._log.debug("auto-scrubbing {0}", util.displayable_path(item.path))
                self._log.debug("{0} will be updated because of {1}", n, model)
        self._log.debug("{} access token: {}", self.data_source, self.access_token)
            self._log.debug("Album removed from Spotify: {}", album_id)
            self._log.debug("Spotify API error: {}", e)
        self._log.debug("Found {} result(s) from {} for '{}'", len(response_data), self.data_source, query)
            self._log.debug("Your beets query returned no items, skipping {}.", self.data_source)
                self._log.debug("{} track(s) found, count: {}", self.data_source, len(response_data_tracks))
                self._log.debug("Most popular track chosen, count: {}", len(response_data_tracks))
        self._log.debug("Total {} tracks", len(items))
                    self._log.debug("Popularity already present for: {}", item)
                self._log.debug("No track_id present for: {}", item)
        self._log.debug("track_popularity: {} and track_isrc: {}", track_data.get("popularity"), track_data.get("external_ids").get("isrc"))
            self._log.debug("Spotify API error: {}", e)
        self._log.debug("URL is {0}", url)
        self._log.debug("auth type is {0}", config["subsonic"]["auth"])
                    self._log.debug('"{0}" -> "{1}"', text, r)
        self._log.debug("using {0.name} to compute URIs", uri_getter)
        self._log.debug("generating thumbnail for {0}", album)
                self._log.debug("found a suitable {1}x{1} thumbnail for {0}, " "forcing regeneration", album, size)
                self._log.debug("{1}x{1} thumbnail for {0} exists and is " "recent enough", album, size)
        self._log.debug("Wrote file {0}", displayable_path(outfilename))
                self._log.debug("{0}: {1} -> None", field, value)
                log.debug("skipping {0} because mtime is up to date ({1})", displayable_path(item.path), item.mtime)
                log.debug("emptied album {0}", album_id)
                log.debug("moving album {0}", album_id)
            log.debug("moving: {0}", util.displayable_path(obj.path))
            log.debug("Invalid color_name: {0}", color_name)
    log.debug("plugin paths: {0}", util.displayable_path(paths))
        log.debug("overlaying configuration: {0}", util.displayable_path(overlay_path))
        log.debug("user configuration: {0}", util.displayable_path(config_path))
        log.debug("no user configuration found at {0}", util.displayable_path(config_path))
    log.debug("data directory: {0}", util.displayable_path(config.config_dir()))
        log.debug("{}", traceback.format_exc())
    log.debug("library database: {0}\n" "library directory: {1}", util.displayable_path(lib.path), util.displayable_path(lib.directory))
        log.debug("{}", traceback.format_exc())
        log.debug("{}", traceback.format_exc())
                    log.debug("ImageMagick version check failed: {}", exc)
        log.debug("artresizer: ImageMagick resizing {0} to {1}", displayable_path(path_in), displayable_path(path_out))
            log.debug("`convert` exited with (status {}) when " "getting size with command {}:\n{}", exc.returncode, cmd, exc.output.strip())
        log.debug("comparing images with pipeline {} | {}", convert_cmd, compare_cmd)
            log.debug("ImageMagick convert failed with status {}: {!r}", convert_proc.returncode, convert_stderr)
                log.debug("ImageMagick compare failed: {0}, {1}", displayable_path(im2), displayable_path(im1))
            log.debug("IM output is not a number: {0!r}", out_str)
        log.debug("ImageMagick compare score: {0}", phash_diff)
        log.debug("artresizer: PIL resizing {0} to {1}", displayable_path(path_in), displayable_path(path_out))
                    log.debug("PIL Pass {0} : Output size: {1}B", i, filesize)
```

See `oneline_python` shell function if you want to check yourself

```fish
oneline_python() {
  # args: <python-file> ...
  # info: _unformat given .py files: remove newlines from each statement
  sed -zr '
    s/(^|\n) *# [^\n]*//g
    s/  # [^\n]*//g
    s/,?\n\s*([]}).]+)/\1/g
    s/\n\s+(\b(and|or)\b|==)/ \1/g
    s/,\s*\n\s+/, /g
    s/"\s*\n\s+(%|f['\''"])/" \1/g
    s/([[{(])\s*\n\s+/\1/g
    s/(["'\''])\n\s*(["'\''+])/\1 \2/g
    s/\n\s*( \+)/\1/g
    s/(\[[^]\n]+)\n\s*( for)/\1\2/g
  ' $@ |
    tr '\000' '\n'
}
```

Reply:
I think it's more important to get the rest of the PR merged than to argue back and forth over this right now.  Let me fix this change, and any other logging-related f-strings I added, and then we should merge.